### ISSUE TITLE
Telemetry: An error has occurred during metrics gathering: label dimensions inconsistent with previously collected metrics in the same metric family

### ISSUE THREAD

**Issue by bobbytables** [istio/istio#8906]
**Describe the bug**
Istio Telemetry starts to fail on `/metrics` endpoint for prometheus metrics

**Expected behavior**
Consistent labels should exist

**Steps to reproduce the bug**
Install istio in a system that emits metrics and wait I guess?

**Version**
Kubernets 1.9 + Istio 1.0

**Installation**
Helm installation with slight modification

**Environment**
AWS w/ Kube-AWS

**Cluster state**
The amount of data the tool exposes about our cluster is unacceptable for me to post here.

Full error:

```
An error has occurred during metrics gathering:

2 error(s) occurred:
* collected metric istio_request_bytes label:<name:"connection_security_policy" value:"unknown" > label:<name:"destination_app" value:"permissions" > label:<name:"destination_principal" value:"unknown" > label:<name:"destination_service" value:"permissions-public.applications.svc.cluster.local" > label:<name:"destination_service_name" value:"permissions-public" > label:<name:"destination_service_namespace" value:"applications" > label:<name:"destination_version" value:"8" > label:<name:"destination_version" value:"8" > label:<name:"destination_workload" value:"permissions-grpc-v008" > label:<name:"reporter" value:"source" > label:<name:"request_protocol" value:"grpc" > label:<name:"response_code" value:"200" > label:<name:"source_app" value:"benefits" > label:<name:"source_principal" value:"unknown" > label:<name:"source_version" value:"unknown" > label:<name:"source_workload" value:"benefits-grpc" > label:<name:"source_workload_namespace" value:"applications" > histogram:<sample_count:105 sample_sum:4515 bucket:<cumulative_count:0 upper_bound:1 > bucket:<cumulative_count:0 upper_bound:10 > bucket:<cumulative_count:105 upper_bound:100 > bucket:<cumulative_count:105 upper_bound:1000 > bucket:<cumulative_count:105 upper_bound:10000 > bucket:<cumulative_count:105 upper_bound:100000 > bucket:<cumulative_count:105 upper_bound:1e+06 > bucket:<cumulative_count:105 upper_bound:1e+07 > bucket:<cumulative_count:105 upper_bound:1e+08 > >  has label dimensions inconsistent with previously collected metrics in the same metric family
* collected metric istio_request_bytes label:<name:"connection_security_policy" value:"unknown" > label:<name:"destination_principal" value:"unknown" > label:<name:"destination_service" value:"sessions-grpc-gateway.applications.svc.cluster.local" > label:<name:"destination_service_name" value:"sessions-grpc-gateway" > label:<name:"destination_service_namespace" value:"applications" > label:<name:"destination_version" value:"unknown" > label:<name:"destination_version" value:"unknown" > label:<name:"destination_workload" value:"sessions" > label:<name:"destination_workload_namespace" value:"applications" > label:<name:"reporter" value:"source" > label:<name:"request_protocol" value:"http" > label:<name:"response_code" value:"200" > label:<name:"source_app" value:"unknown" > label:<name:"source_principal" value:"unknown" > label:<name:"source_version" value:"unknown" > label:<name:"source_workload" value:"sessions" > label:<name:"source_workload_namespace" value:"applications" > histogram:<sample_count:29 sample_sum:275 bucket:<cumulative_count:28 upper_bound:1 > bucket:<cumulative_count:28 upper_bound:10 > bucket:<cumulative_count:28 upper_bound:100 > bucket:<cumulative_count:29 upper_bound:1000 > bucket:<cumulative_count:29 upper_bound:10000 > bucket:<cumulative_count:29 upper_bound:100000 > bucket:<cumulative_count:29 upper_bound:1e+06 > bucket:<cumulative_count:29 upper_bound:1e+07 > bucket:<cumulative_count:29 upper_bound:1e+08 > >  has label dimensions inconsistent with previously collected metrics in the same metric family
```

**Comment by douglas-reid** [comment]
The labels being generated are definitely different. Do you have two differing configs for `istio_requests_total` in your prometheus adapter config OR do you have two different `requestcount.metric` definitions deployed to the API server?

**Comment by douglas-reid** [comment]
There is other weirdness here too:

`label:<name:"destination_version" value:"8" > label:<name:"destination_version" value:"8" >`

Why is `destination_version` in the labelset twice?  Do you have config that is attempting to do that?

**Comment by bobbytables** [comment]
> There is other weirdness here too:
> 
> `label:<name:"destination_version" value:"8" > label:<name:"destination_version" value:"8" >`
> 
> Why is `destination_version` in the labelset twice? Do you have config that is attempting to do that?

Nope, this is a fresh install of Istio actually:

```
➜  ~ int20 get handlers.config.istio.io --all-namespaces
No resources found.
➜  ~ int20 get metrics.config.istio.io --all-namespaces
NAMESPACE      NAME              AGE
istio-system   requestcount      3h
istio-system   requestduration   3h
istio-system   requestsize       3h
istio-system   responsesize      3h
istio-system   tcpbytereceived   3h
istio-system   tcpbytesent       3h
➜  ~ int20 get prometheuses.config.istio.io --all-namespaces
NAMESPACE      NAME      AGE
istio-system   handler   3h
```

**Comment by bobbytables** [comment]
Here's how I got here: I saw our metrics in datadog stopped being reported (we're using the DataDog istio integration). I started to investigate and noticed that the istio-telemetry pod was throwing `500`s for all requests.

I port forwarded the single istio-telemetry pod locally using `kubectl port-forward` and grabbed the content and posted it in this ticket. I haven't done anything else.

**Comment by douglas-reid** [comment]
@bobbytables which version of Istio are you testing on?

@mandarjog have there been any changes to any of the instance creation code, etc., for the gRPC work that could be creating memory corruption in instances or config?

In the example above, the instances sent to the `prometheus` adapter has duplicate labels (
`destination_version`) and one is missing a label `destination_app`.

**Comment by bobbytables** [comment]
@douglas-reid v1.0.0 - also. Just refreshed telemetry output and a new metric popped in (from 2 to 3 errors in prometheus now)

```
3 error(s) occurred:
* collected metric istio_response_bytes label:<name:"connection_security_policy" value:"none" > label:<name:"destination_app" value:"payrollsteps" > label:<name:"destination_app" value:"payrollsteps" > label:<name:"destination_app" value:"payrollsteps" > label:<name:"destination_principal" value:"unknown" > label:<name:"destination_service" value:"payrollsteps-public.applications.svc.cluster.local" > label:<name:"destination_service_name" value:"payrollsteps-public" > label:<name:"destination_service_namespace" value:"applications" > label:<name:"destination_workload" value:"payrollsteps-public" > label:<name:"destination_workload_namespace" value:"applications" > label:<name:"request_protocol" value:"http" > label:<name:"response_code" value:"422" > label:<name:"source_app" value:"unknown" > label:<name:"source_principal" value:"unknown" > label:<name:"source_version" value:"unknown" > label:<name:"source_workload" value:"unknown" > label:<name:"source_workload_namespace" value:"unknown" > histogram:<sample_count:1 sample_sum:82 bucket:<cumulative_count:0 upper_bound:1 > bucket:<cumulative_count:0 upper_bound:10 > bucket:<cumulative_count:1 upper_bound:100 > bucket:<cumulative_count:1 upper_bound:1000 > bucket:<cumulative_count:1 upper_bound:10000 > bucket:<cumulative_count:1 upper_bound:100000 > bucket:<cumulative_count:1 upper_bound:1e+06 > bucket:<cumulative_count:1 upper_bound:1e+07 > bucket:<cumulative_count:1 upper_bound:1e+08 > >  has label dimensions inconsistent with previously collected metrics in the same metric family
* collected metric istio_request_bytes label:<name:"connection_security_policy" value:"unknown" > label:<name:"destination_app" value:"permissions" > label:<name:"destination_principal" value:"unknown" > label:<name:"destination_service" value:"permissions-public.applications.svc.cluster.local" > label:<name:"destination_service_name" value:"permissions-public" > label:<name:"destination_service_namespace" value:"applications" > label:<name:"destination_version" value:"8" > label:<name:"destination_version" value:"8" > label:<name:"destination_workload" value:"permissions-grpc-v008" > label:<name:"reporter" value:"source" > label:<name:"request_protocol" value:"grpc" > label:<name:"response_code" value:"200" > label:<name:"source_app" value:"benefits" > label:<name:"source_principal" value:"unknown" > label:<name:"source_version" value:"unknown" > label:<name:"source_workload" value:"benefits-grpc" > label:<name:"source_workload_namespace" value:"applications" > histogram:<sample_count:106 sample_sum:4558 bucket:<cumulative_count:0 upper_bound:1 > bucket:<cumulative_count:0 upper_bound:10 > bucket:<cumulative_count:106 upper_bound:100 > bucket:<cumulative_count:106 upper_bound:1000 > bucket:<cumulative_count:106 upper_bound:10000 > bucket:<cumulative_count:106 upper_bound:100000 > bucket:<cumulative_count:106 upper_bound:1e+06 > bucket:<cumulative_count:106 upper_bound:1e+07 > bucket:<cumulative_count:106 upper_bound:1e+08 > >  has label dimensions inconsistent with previously collected metrics in the same metric family
* collected metric istio_request_bytes label:<name:"connection_security_policy" value:"unknown" > label:<name:"destination_principal" value:"unknown" > label:<name:"destination_service" value:"sessions-grpc-gateway.applications.svc.cluster.local" > label:<name:"destination_service_name" value:"sessions-grpc-gateway" > label:<name:"destination_service_namespace" value:"applications" > label:<name:"destination_version" value:"unknown" > label:<name:"destination_version" value:"unknown" > label:<name:"destination_workload" value:"sessions" > label:<name:"destination_workload_namespace" value:"applications" > label:<name:"reporter" value:"source" > label:<name:"request_protocol" value:"http" > label:<name:"response_code" value:"200" > label:<name:"source_app" value:"unknown" > label:<name:"source_principal" value:"unknown" > label:<name:"source_version" value:"unknown" > label:<name:"source_workload" value:"sessions" > label:<name:"source_workload_namespace" value:"applications" > histogram:<sample_count:29 sample_sum:275 bucket:<cumulative_count:28 upper_bound:1 > bucket:<cumulative_count:28 upper_bound:10 > bucket:<cumulative_count:28 upper_bound:100 > bucket:<cumulative_count:29 upper_bound:1000 > bucket:<cumulative_count:29 upper_bound:10000 > bucket:<cumulative_count:29 upper_bound:100000 > bucket:<cumulative_count:29 upper_bound:1e+06 > bucket:<cumulative_count:29 upper_bound:1e+07 > bucket:<cumulative_count:29 upper_bound:1e+08 > >  has label dimensions inconsistent with previously collected metrics in the same metric family
```

**Comment by bobbytables** [comment]
I just deleted the istio-system namespace and redeployed with some config updates (this is an integration cluster internally for our org) today, but we've been running 1.0 on this cluster since it dropped a few months ago.

However, this update was to get metrics into DataDog so I turned on telemetry pod.

https://gist.github.com/bobbytables/2a22f819e05f66ecec5a553b7a52f0e3

**Comment by bobbytables** [comment]
Looks like the prometheus handler isn't even defining this metric we're receiving looking closely.

**Comment by douglas-reid** [comment]
@bobbytables correct. the prometheus handler is saying which labels should be there, but the instance is defined by the metric config. both of those look fine from a quick glance, which means something else is wonky. i was hoping that there would be some sort of mangled config to make debugging this easier.

**Comment by bobbytables** [comment]
Same :)

I'm wondering if this cluster has a demon of some sort from previous versions of istio that we've installed on it (This cluster has had 0.5 to 1.0 on it)

**Comment by douglas-reid** [comment]
@bobbytables i never close my mind to the possibilities of demons wrecking havoc on deployments. Especially given how wonky the upgrade story has been in those generations.

**Comment by douglas-reid** [comment]
@bobbytables I hesitate to ask this, but... what happens if you kill the istio-telemetry pod? Do things get better?

**Comment by bobbytables** [comment]
Yup, just deleted it and all of the metrics are back.

**Comment by bobbytables** [comment]
I imagine something somwhere will post a metric that is invalid and it'll break soon.

**Comment by bobbytables** [comment]
I think Telemetry has a bug with its consistency, the errors are variable in number. Here's a vid to show you what I mean.

https://youtu.be/Qniiv6H2Tis

**Comment by mandarjog** [comment]
Can you post telemetry logs? What surprises me is that this happens on a fresh install.
@douglas-reid grpc should not (tm) affect the static flow.

**Comment by douglas-reid** [comment]
@mandarjog if this wasn't impacted by work done for grpc, I'm concerned that the (newer) runtime has some issue in the instance creation logic. There was a second report of something similar happening recently on the mailing-list. It looks, superficially, like memory reuse issues or related.

**Comment by douglas-reid** [comment]
@bobbytables that video could perhaps be explained by different `Report` batches being sent to Mixer, I suppose. I don't know what your traffic looks like. I believe however that isn't the root cause of this issue.

**Comment by bobbytables** [comment]
@douglas-reid @mandarjog - I think there might be a race or something in Telemetry, as we noticed, the metrics in the _error_ output has duplicate labels in it. So here's what I did.

1. Bounced telemetry pod which revived it.
1. Port-forwarded new telemetry pod and grabbed `/metrics` output
1. Waited for it to fail and grabbed the error output
1. Compared the erroring metric labels to old labels from step 2

Here's the result

# Original Metric Labels (When Telemetry is working just fine)

```
connection_security_policy="none"
destination_app="payrollsteps"
destination_principal="unknown"
destination_service="payrollsteps-public.applications.svc.cluster.local"
destination_service_name="payrollsteps-public"
destination_service_namespace="applications"
destination_version="unknown"
destination_workload="payrollsteps-public"
destination_workload_namespace="applications"
reporter="destination"
request_protocol="http"
response_code="200"
source_app="unknown"
source_principal="unknown"
source_version="unknown"
source_workload="unknown"
source_workload_namespace="unknown"
```

# New Metric Labels (Failing Metric in Prometheus Endpoint)

```
label:<name:"connection_security_policy" value:"none" <---------- Duplicate!
label:<name:"connection_security_policy" value:"none"
label:<name:"destination_app" value:"payrollsteps" 
label:<name:"destination_app" value:"payrollsteps" <---------- Duplicate!
label:<name:"destination_principal" value:"unknown"
label:<name:"destination_service" value:"payrollsteps-public.applications.svc.cluster.local"
label:<name:"destination_service_name" value:"payrollsteps-public"
label:<name:"destination_service_namespace" value:"applications"
label:<name:"destination_version" value:"unknown"
label:<name:"destination_workload_namespace" value:"applications"
label:<name:"request_protocol" value:"http"
label:<name:"response_code" value:"302"
label:<name:"source_app" value:"unknown"
label:<name:"source_principal" value:"unknown"
label:<name:"source_version" value:"unknown"
label:<name:"source_workload" value:"unknown"
label:<name:"source_workload_namespace" value:"unknown"
histogram:<sample_count:3 sample_sum:0 bucket:<cumulative_count:3 upper_bound:1
```

This smells like a race condition or something that isn't maintaining dimensions correctly in Istio.

**Comment by douglas-reid** [comment]
@bobbytables do you have another cluster that you can replicate this issue in? i'm not able to repro with any of my long term clusters. Also, when you have the errors, do you have the corresponding `stdio` logs from the telemetry pod?  Do those show duplication in labels?

**Comment by bobbytables** [comment]
I turned on debug level logs for `telemetry` and it doesn't include the attribute bag details:

It just says it's dispatching and that it's done

```
2018-09-25T17:07:27.149825Z	debug	begin dispatch: destination='kubernetes:handler.kubernetesenv.istio-system(kubernetesenv)'
2018-09-25T17:07:27.149897Z	debug	complete dispatch: destination='kubernetes:handler.kubernetesenv.istio-system(kubernetesenv)' {err:<nil>}
```

**Comment by douglas-reid** [comment]
Do you have any logs generated by the `stdio` for the same request?  They should look something like:

`{"sourceIp":"10.36.2.8","requestId":"aba5a631-6f95-93c0-b422-4a1d78368f24","sourceWorkload":"istio-ingressgateway","clientTraceId":"","reporter":"source","destinationIp":"10.36.1.7","httpAuthority":"35.199.189.190:15031","apiClaims":"","latency":"3.584311ms","level":"info","receivedBytes":836,"desti…`

**Comment by douglas-reid** [comment]
@bobbytables apologies. I believed these issues were being seen in mixer dispatch. A closer look reveals that these are being generated in the prometheus client library during a `Gather()`. This is likely a different issue -- and digging is still needed to figure out what could be causing that issue.

**Comment by douglas-reid** [comment]
I've posted a [question to prometheus-users](https://groups.google.com/d/msg/prometheus-users/QnP2f8NdpZc/VWRqupKkAQAJ) about this issue to see if we can get some additional help in troubleshooting.

**Comment by bobbytables** [comment]
@beorn7 - Neat. If this week has enough hours for me to do that (point istio at tip) we can do a build of Istio and install it in our lower level cluster.

**Comment by beorn7** [comment]
I have just merged https://github.com/prometheus/client_golang/pull/474 . With that one in, you should get the correct error message if your case still occurs with the current state of client_golang.

**Comment by douglas-reid** [comment]
@bobbytables any update on replication with the latest client_golang?

**Comment by beorn7** [comment]
Yeah, I would be very interested in the results of that...

**Comment by beorn7** [comment]
https://github.com/prometheus/client_golang/pull/478 suggests current client_golang is working fine.

**Comment by michaelkipper** [comment]
I'm seeing this too, in a close-to-stock installation of `1.0.2` in our staging cluster.
More info here: https://groups.google.com/d/msg/istio-users/lsdhM2meMhM/Wwx7lCAXBAAJ

**Comment by corey-aloia-sap** [comment]
Any update on this thread? We are experiencing this with Istio 1.0.1. When I delete the telemetry pod, it will then work for ~1 hour before hitting this again. We were using Istio 0.8, and build our dashboards based on this. In 0.8 we never ran into any issues. Any update on this would be great as its blocking us :)

**Comment by sebarys** [comment]
Also facing this issue with Istio 1.0.1.
Is there any update or any way to mitigate this?

**Comment by mhamrah** [comment]
This can be closed, we updated to 1.0.4 and have seen stable telemetry metrics.

**Comment by bobbytables** [comment]
Nope, still seeing this:

```
* collected metric "istio_requests_total" { label:<name:"connection_security_policy" value:"unknown" > label:<name:"destination_app" value:"locator" > label:<name:"destination_service" value:"locator.applications.svc.cluster.local" > label:<name:"destination_service_name" value:"locator" > label:<name:"destination_service_namespace" value:"applications" > label:<name:"destination_version" value:"unknown" > label:<name:"destination_version" value:"unknown" > label:<name:"destination_workload" value:"locator-grpc" > label:<name:"destination_workload_namespace" value:"applications" > label:<name:"reporter" value:"source" > label:<name:"request_protocol" value:"grpc" > label:<name:"response_code" value:"200" > label:<name:"source_app" value:"hcm-web" > label:<name:"source_principal" value:"unknown" > label:<name:"source_version" value:"unknown" > label:<name:"source_workload" value:"hcm-web-public" > label:<name:"source_workload_namespace" value:"applications" > counter:<value:8 > } has two or more labels with the same name: destination_version
```

**Comment by douglas-reid** [comment]
For those following along, I mentioned this on the related issue, but here it is again: https://github.com/prometheus/client_golang/pull/511

**Comment by 01deyishu** [comment]
same issue with istio-1.0.4 + kubernetes 1.9

**Comment by douglas-reid** [comment]
This issue was root-caused and addressed in the 1.0 and 1.1 branches. When 1.0.6 and 1.1 are out, the fix will be to upgrade.
